src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs

```diff
@@ -72,18 +72,14 @@ private static IArgumentConstraint CreateEqualityConstraint(object expressionVal
             return new EqualityArgumentConstraint(expressionValue);
         }
 
-        private static object InvokeExpression(Expression expression)
-        {
-            return Expression.Lambda(expression).Compile().DynamicInvoke();
-        }
-
         private IArgumentConstraint GetArgumentConstraintFromExpression(Expression expression)
         {
             object expressionValue = null;
 
             var trappedConstraints = this.argumentConstraintTrapper.TrapConstraints(() =>
             {
-                expressionValue = InvokeExpression(expression);
+                // in this ParsedArgumentExpression scenario, we get benefits from the same optimizations as ParsedCallExpression
+                expressionValue = ParsedCallExpression.FastEvaluate(expression);

```

Comment:
I convinced myself it would work, here is the argument I made.

The input to InvokeExpression is an expression tree. The requirement for correctness is that FastEvaluate evaluates the expression tree equivalently to Lambda(...).Compile().DynamicEvoke() with respect to a) results and b) side effects.

But in short: it doesn't really matter how the expression tree is evaluated as long as it is consistent with normal C# semantics. Claim: FastEvaluate is consistent with normal C# semantics.

This is easy to demonstrate because there we are definitely not altering *order of evaluation* of property getter, field getter, and constant terms because FastEvaluate never actually makes any choice about the order to evaluate in: it is a straight-line path, and so ultimately, Lambda(...).Compile() has to compile down to an equivalent sequence of get operations, which have the same side effects because they call the same field/property getters in the same order.